crypto/evp/evp_local.h

```diff
@@ -65,6 +66,43 @@ struct evp_kdf_ctx_st {
     void *data;                 /* Algorithm-specific data */
 } /* EVP_KDF_CTX */ ;
 
+struct evp_rand_ctx_st {
+    EVP_RAND *meth;             /* Method structure */
+    void *data;                 /* Algorithm-specific data */
+    size_t max_request;         /*
+                                 * Cached: maximum number of bytes generated 
+                                 * in a single call to the generate function
+                                 */
+    unsigned int strength;      /* Cache: bit strength of generator */
+} /* EVP_RAND_CTX */ ;
+
+struct evp_rand_st {
+    OSSL_PROVIDER *prov;
+    int name_id;
+    CRYPTO_REF_COUNT refcnt;
+    CRYPTO_RWLOCK *lock;
```

Comment:
This rwlock is used solely for `CRYPTO_{UP,DOWN}_REF`. If you would rename it to `refcnt_lock`, then you could remove the awkward "prov_" prefix from the "prov_lock"-functions below:

```C
    CRYPTO_REF_COUNT refcnt;
    CRYPTO_RWLOCK *refcnt_lock;
   ...

    OSSL_OP_rand_enable_locking_fn *enable_locking;
    OSSL_OP_rand_lock_fn *lock;
    OSSL_OP_rand_unlock_fn *unlock;
```

There is a precedent `prov->refcnt_lock` in provider_core.c for using this name. In fact, the source code contains many `lock` members which are used as reference count locks only (i.e. for `CRYPTO_{UP,DOWN}_REF(&xxx->refcnt, &ref, xxx->lock)` ) and which would be more properly named `refcnt_lock`. But that's an issue for a different pr.

Reply:
Okay.

For the future: we should wrapper up the reference counting and locking into a structure and not even allocate the lock when it isn't required (i.e. just about all of the time).

Reply:
> not even allocate the lock when it isn't required (i.e. just about all of the time).

Are you thinking the case where `HAVE_ATOMICS` is defined?

https://github.com/openssl/openssl/blob/24b6261e5bb286fa494e7208a3de28365e0ca004/include/internal/refcount.h#L29

I had similar thoughts yesterday too. It's quite a waste of resources to allocate all these unused rwlocks.

Reply:
I thought it was slightly more general that that, but I'm likely misremembering.